File: backend-dotnet/FinanceTracker.Api/Contracts/ReportDtos.cs
namespace FinanceTracker.Api.Contracts;

public sealed record CategorySpendItem(
    string category,
    decimal amount
);

public sealed record CategorySpendResponse(
    IReadOnlyList<CategorySpendItem> items,
    decimal total
);

public sealed record IncomeExpenseTrendItem(
    DateOnly date,
    decimal income,
    decimal expense
);

public sealed record Point(
    DateOnly date,
    decimal balance
);

public sealed record AccountBalanceTrendItem(
    string accountName,
    IReadOnlyList<Point> points
);

public sealed record TrendPoint(
    DateOnly date,
    decimal amount
);

public sealed record InsightItem(
    string type,
    string title,
    string description,
    string severity
);

public sealed record FutureBalancePredictionResponse(
    decimal currentBalance,
    decimal projectedRecurringNet,
    decimal averageDailySpending,
    decimal predictedBalance,
    int horizonDays
);

public sealed record SavingsTrendPoint(
    DateOnly date,
    decimal income,
    decimal expense,
    decimal savings
);

public sealed record CategoryTrendSeries(
    string category,
    IReadOnlyList<TrendPoint> points
);

public sealed record TrendsResponse(
    IReadOnlyList<IncomeExpenseTrendItem> incomeExpense,
    IReadOnlyList<SavingsTrendPoint> savings,
    IReadOnlyList<CategoryTrendSeries> categoryTrends
);

public sealed record NetWorthPoint(
    DateOnly date,
    decimal netWorth
);

public sealed record NetWorthResponse(
    decimal currentNetWorth,
    IReadOnlyList<NetWorthPoint> points
);

File: backend/src/main/java/com/finance/dto/ReportDtos.java
package com.finance.dto;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;
import java.util.Map;

public class ReportDtos {

    public record CategorySpendResponse(
            List<Map<String, Object>> items,
            BigDecimal total
    ) {
    }

    public record IncomeExpenseTrendItem(
            LocalDate date,
            BigDecimal income,
            BigDecimal expense
    ) {
    }

    public record AccountBalanceTrendItem(
            String accountName,
            List<Point> points
    ) {
    }

    public record Point(
            LocalDate date,
            BigDecimal balance
    ) {
    }

    public record InsightItem(
            String type,
            String title,
            String description,
            String severity
    ) {
    }

    public record FutureBalancePredictionResponse(
            BigDecimal currentBalance,
            BigDecimal projectedRecurringNet,
            BigDecimal averageDailySpending,
            BigDecimal predictedBalance,
            Integer horizonDays
    ) {
    }
}
